sktime/datasets/_dataset_downloader.py

```diff
@@ -0,0 +1,248 @@
+"""Utilities for downloading datasets."""
+
+__author__ = ["jgyasu"]
+
+__all__ = ["DatasetDownloader"]
+
+import os
+import shutil
+import tempfile
+import zipfile
+from pathlib import Path
+from urllib.error import HTTPError, URLError
+from urllib.request import urlretrieve
+
+from sktime.utils.dependencies import _safe_import
+
+snapshot_download = _safe_import(
+    "huggingface_hub.snapshot_download", pkg_name="huggingface-hub"
+)
+hf_hub_download = _safe_import(
+    "huggingface_hub.hf_hub_download", pkg_name="huggingface-hub"
+)
+HfHubHTTPError = _safe_import(
+    "huggingface_hub.utils.HfHubHTTPError", pkg_name="huggingface-hub"
+)
+RepositoryNotFoundError = _safe_import(
+    "huggingface_hub.utils.RepositoryNotFoundError", pkg_name="huggingface-hub"
+)
+
+
+class DatasetDownloadStrategy:
+    """Base class for dataset download strategies."""
+
+    def download_dataset(self, **kwargs):
+        """Download a dataset using this strategy."""
+        raise NotImplementedError("Subclasses must implement download_dataset method")
+
+
+class HuggingFaceDownloader(DatasetDownloadStrategy):
+    """Download datasets using the Hugging Face Hub snapshot mechanism.
+
+    Implements a strategy to download datasets stored in Hugging Face
+    repositories.
+
+    Parameters
+    ----------
+    repo_name : str
+        Name of the Hugging Face repository containing the datasets.
+        In the format: "organisation/repository_name",
+        e.g., "sktime/tsf-datasets"
+    repo_type : str, default="dataset"
+        Type of repository (e.g., "dataset", "model", etc.).
+
+    References
+    ----------
+    https://huggingface.co/docs/huggingface_hub/en/guides/download#filter-files-to-download
+    """
+
+    def __init__(self, repo_name, repo_type="dataset"):
+        self.repo_name = repo_name
+        self.repo_type = repo_type
+
+    def download_dataset(
+        self,
+        dataset_folder,
+        download_path=None,
+        force_download=False,
+        token=None,
+        **kwargs,
+    ):
+        """Download a specific dataset folder from the Hugging Face repository.
+
+        Parameters
+        ----------
+        dataset_folder : str
+            Name of the folder (dataset) inside the repository to download.
+        download_path : str or Path, optional
+            Local directory where the dataset should be saved. If not specified,
+            defaults to a 'local_data' folder in the current working directory.
+        force_download : bool, default=False
+            Whether to force re-download the dataset even if it's cached locally.
+        token : str, optional
```

Comment:
should the token not be part of the `__init__` params?

If we use the strategy pattern, the args of `download_dataset` should be the same across all strategies.